plugins/uefi-capsule/fu-uefi-bootmgr.c

```diff
@@ -292,6 +293,165 @@ fu_uefi_copy_asset (const gchar *source, const gchar *target, GError **error)
 	return TRUE;
 }
 
+static gboolean
+fu_uefi_bootmgr_setup_grub (FuDevice *device,
+			    const gchar *esp_path,
+			    const gchar *target_app,
+			    GError **error)
+{
+	g_autofree gchar *grub_reboot = NULL;
+	g_autofree gchar *grub_target = NULL;
+	g_autofree gchar *sysconfdir = NULL;
+
+	/* verify GRUB snippet exists */
+	sysconfdir = fu_common_get_path (FU_PATH_KIND_SYSCONFDIR);
+	grub_target = g_build_filename (sysconfdir, "grub.d", "35_fwupd", NULL);
+	if (!g_file_test (grub_target, G_FILE_TEST_EXISTS)) {
+		g_autofree gchar *buf = NULL;
+		g_autofree gchar *datadir = NULL;
+		g_autofree gchar *grub_source = NULL;
+		g_autofree gchar *mount = NULL;
+		g_autofree gchar *update_grub = NULL;
+		g_autoptr(GString) str = NULL;
+
+		/* read config snippet */
+		datadir = fu_common_get_path (FU_PATH_KIND_DATADIR_PKG);
+		grub_source = g_build_filename (datadir, "fwupd.grub.conf", NULL);
+		if (!g_file_get_contents (grub_source, &buf, NULL, error))
+			return FALSE;
+
+		/* replace ESP path with what we detected */
+		str = g_string_new (buf);
+		fu_common_string_replace (str, "##EFIPATH##", target_app);
+		fu_common_string_replace (str, esp_path, "");
+		fu_common_string_replace (str, "##ESP##", esp_path);
+		if (!g_file_set_contents (grub_target, str->str, -1, error))
+			return FALSE;
+		if (chmod (grub_target, 0755) == -1) {
+			g_set_error (error,
+				     G_IO_ERROR,
+#ifdef HAVE_ERRNO_H
+				     g_io_error_from_errno (errno),
+#else
+				     G_IO_ERROR_FAILED,
+#endif
+				     "chmod failed for %s", grub_target);
+			return FALSE;
+		}
+
+		/* refresh GRUB configuration */
+//Fedora needs
+// grub2-mkconfig -o "$(readlink -e /etc/grub2.conf)"
+//IIUC
+		update_grub = fu_common_find_program_in_path ("update-grub", NULL);
+		if (update_grub != NULL) {
+			g_autofree gchar *output = NULL;
+			if (!g_spawn_command_line_sync (update_grub,
+							&output, NULL, NULL, error))
+				return FALSE;
+			g_debug ("%s", output);
+		} else {
+			g_warning ("unable to refresh GRUB configuration");
+		}
+	}
+	grub_reboot = fu_common_find_program_in_path ("grub-reboot", NULL);
+	if (grub_reboot == NULL)
+		grub_reboot = fu_common_find_program_in_path ("grub2-reboot", NULL);
+	if (grub_reboot != NULL) {
+		g_autofree gchar *cmd = g_strdup_printf ("%s fwupd", grub_reboot);
```

Comment:
@martinezjavier can we do this by label?

Reply:
Yes, either menu entry id or label work.